src/rust/engine/src/context.rs

```diff
@@ -229,87 +324,63 @@ impl Core {
-            // We unwrap because global_options.py will have already validated these are defined.
-            remoting_opts.execution_address.as_ref().unwrap(),
-            remoting_opts.store_address.as_ref().unwrap(),
-            process_execution_metadata.clone(),
-            root_ca_certs.clone(),
-            remoting_opts.execution_headers.clone(),
-            full_store.clone(),
-            // TODO if we ever want to configure the remote platform to be something else we
-            // need to take an option all the way down here and into the remote::CommandRunner struct.
-            Platform::Linux_x86_64,
-            remoting_opts.execution_overall_deadline,
-            Duration::from_millis(100),
-            remoting_opts.execution_rpc_concurrency,
-            remoting_opts.cache_rpc_concurrency,
-            remoting_opts.cache_read_timeout,
-            capabilities_cell_opt,
-          )?),
-          exec_strategy_opts.remote_parallelism,
-        ))
-      } else if remote_caching_used {
-        Box::new(remote_cache::CommandRunner::new(
-          local_command_runner.into(),
-          process_execution_metadata.clone(),
-          executor.clone(),
-          full_store.clone(),
-          remote_store_address.as_ref().unwrap(),
-          root_ca_certs.clone(),
-          remoting_opts.store_headers.clone(),
-          Platform::current()?,
-          exec_strategy_opts.remote_cache_read,
-          exec_strategy_opts.remote_cache_write,
-          remoting_opts.cache_warnings_behavior,
-          remoting_opts.cache_eager_fetch,
-          remoting_opts.cache_rpc_concurrency,
-          remoting_opts.cache_read_timeout,
+          process_execution_metadata,
+          root_ca_certs,
+          exec_strategy_opts,
+          remoting_opts,
         )?)
       } else {
-        local_command_runner
+        None
       };
 
-    // Possibly use the local cache runner, regardless of remote execution/caching.
-    let maybe_local_cached_command_runner = if exec_strategy_opts.local_cache {
-      Box::new(process_execution::cache::CommandRunner::new(
-        maybe_remote_enabled_command_runner.into(),
-        local_cache.clone(),
-        full_store.clone(),
-        process_execution_metadata.clone(),
+    let maybe_local_cached_runner = if exec_strategy_opts.local_cache {
+      Some(Self::make_local_cached_runner(
+        maybe_remote_cached_runner
+          .clone()
+          .unwrap_or_else(|| leaf_runner.clone()),
+        full_store,
+        local_cache,
+        process_execution_metadata,
       ))
     } else {
-      maybe_remote_enabled_command_runner
+      None
     };
 
-    Ok(maybe_local_cached_command_runner)
+    Ok(
+      vec![
+        // Use all enabled caches on the first attempt.
+        maybe_local_cached_runner,
+        // Remove local caching on the second attempt.
+        maybe_remote_cached_runner,
+        // Remove all caching on the third attempt.
+        Some(leaf_runner),
```

Comment:
I don't think I fully understand this. The same `leaf_runner` will be used multiple times right? But because we use `Arc`, it's the same value every time? If so, maybe highlighting the role of `Arc` in a comment would help

Reply:
That's correct.

Reply:
Okay, I do think this would be helpful to explicitly say in a comment. Totally fine in a follow up PR